File: hello-nodejs/src/client/core/UploadManager.js
/**
 * Upload manager for sending files to the server
 */

import { FileManager } from "./FileManager.js";
import { StreamIdGenerator } from "../util/StreamIdGenerator.js";

export class UploadManager {
  constructor(webSocketClient) {
    this.ws = webSocketClient;
    this.fileManager = new FileManager();
    this.streamIdGenerator = new StreamIdGenerator();
  }

  /**
   * Upload a file to the server.
   *
   * @param {string} filePath - Path to the file
   * @param {number} fileSize - Size of the file
   * @param {Function} progressCallback - Optional progress callback
   * @returns {Promise<string>} Stream ID
   */
  async upload(filePath, fileSize, progressCallback = null) {
    // Generate unique stream ID
    const streamId = this.streamIdGenerator.generate();

    // Send START message
    await this.ws.sendControlMessage({ type: "START", streamId });

    // Wait for STARTED
    const startResponse = await this.ws.receiveControlMessage();
    if (startResponse.type !== "STARTED") {
      throw new Error(`Unexpected response to START: ${startResponse.type}`);
    }

    // Upload file in chunks (8KB to avoid WebSocket fragmentation)
    const uploadChunkSize = 8192;
    let offset = 0;
    let bytesSent = 0;

    while (offset < fileSize) {
      const chunkSize = Math.min(uploadChunkSize, fileSize - offset);
      const chunk = await this.fileManager.readChunk(
        filePath,
        offset,
        chunkSize,
      );

      await this.ws.sendBinary(chunk);

      offset += chunk.length;
      bytesSent += chunk.length;

      // Report progress
      if (progressCallback) {
        const progress = Math.floor((bytesSent * 100) / fileSize);
        progressCallback(bytesSent, fileSize, progress);
      }
    }

    // Send STOP message
    await this.ws.sendControlMessage({ type: "STOP", streamId });

    // Wait for STOPPED
    const stopResponse = await this.ws.receiveControlMessage();
    if (stopResponse.type !== "STOPPED") {
      throw new Error(`Unexpected response to STOP: ${stopResponse.type}`);
    }

    return streamId;
  }
}

File: hello-go/src/client/core/upload_manager.go
package core

import (
	"fmt"

	"github.com/feuyeux/hello-mmap/hello-go/src/client/util"
	"github.com/feuyeux/hello-mmap/hello-go/src/logger"
)

func Upload(ws *WebSocketClient, filePath string, fileSize int64) (string, error) {
	// Generate unique stream ID
	streamID := util.GenerateStreamID()
	logger.Info(fmt.Sprintf("Generated stream ID: %s", streamID))

	// Send START message
	err := ws.SendControlMessage(ControlMessage{
		Type:     "START",
		StreamID: streamID,
	})
	if err != nil {
		return "", fmt.Errorf("failed to send START message: %w", err)
	}

	// Wait for START_ACK
	response, err := ws.ReceiveControlMessage()
	if err != nil {
		return "", fmt.Errorf("failed to receive START_ACK: %w", err)
	}
	if response.Type != "STARTED" {
		return "", fmt.Errorf("unexpected response to START: %s", response.Type)
	}

	// Upload file in chunks
	// Use smaller chunk size (8KB) to avoid WebSocket frame fragmentation
	// which the Java server doesn't handle properly
	const uploadChunkSize = 8192
	var offset int64 = 0
	var bytesSent int64 = 0
	lastProgress := 0

	for offset < fileSize {
		chunkSize := int(Min(int64(uploadChunkSize), fileSize-offset))
		chunk, err := ReadChunk(filePath, offset, chunkSize)
		if err != nil {
			return "", fmt.Errorf("failed to read chunk: %w", err)
		}

		if err := ws.SendBinary(chunk); err != nil {
			return "", fmt.Errorf("failed to send chunk: %w", err)
		}

		offset += int64(len(chunk))
		bytesSent += int64(len(chunk))

		// Report progress
		progress := int(bytesSent * 100 / fileSize)
		if progress >= lastProgress+25 && progress <= 100 {
			logger.Info(fmt.Sprintf("Upload progress: %d/%d bytes (%d%%)", bytesSent, fileSize, progress))
			lastProgress = progress
		}
	}

	// Ensure 100% is reported
	if lastProgress < 100 {
		logger.Info(fmt.Sprintf("Upload progress: %d/%d bytes (100%%)", fileSize, fileSize))
	}

	// Send STOP message
	err = ws.SendControlMessage(ControlMessage{
		Type:     "STOP",
		StreamID: streamID,
	})
	if err != nil {
		return "", fmt.Errorf("failed to send STOP message: %w", err)
	}

	// Wait for STOPPED
	response, err = ws.ReceiveControlMessage()
	if err != nil {
		return "", fmt.Errorf("failed to receive STOPPED: %w", err)
	}
	if response.Type != "STOPPED" {
		return "", fmt.Errorf("unexpected response to STOP: %s", response.Type)
	}

	return streamID, nil
}
